### ISSUE TITLE
Suggestion for the migration lock

### ISSUE THREAD

**Issue by electrotype** [tgriesser/knex#2502]
It would be useful if a timestamp column would be added to the migration `xxx_lock` table. That would allow a custom logic to reset the lock when a process crashes and the lock has been taken for too long.

**Comment by elhigu** [comment]
Yes I've been thinking about this myself too. It would also allow us to define maximum locking time in configuration.

**Comment by electrotype** [comment]
I found some kind of "workaround". In fact, I think I will even call this a "solution"! 

I simply alter the migration lock table by myself, in the _very first_ migration file (This version is for SQL Server):

```typescript
const createMigrationLockTimestamps = async (knexClient: knex) => {
  await knexClient.schema.table(TABLE_NAME_MIGRATIONS_LOCK, table => {
    table
      .specificType(LOCK_COL_NAME_LAST_MODIFIED, 'datetimeoffset(3)')
      .notNullable()
      .defaultTo(knexClient.fn.now());
  });

  await knexClient.raw(`
    CREATE TRIGGER lockModified
    ON dbo.${TABLE_NAME_MIGRATIONS_LOCK}
    AFTER UPDATE 
    AS
      UPDATE dbo.${TABLE_NAME_MIGRATIONS_LOCK}
      SET ${LOCK_COL_NAME_LAST_MODIFIED} = SYSDATETIME()
  `);
};
```
I can then manage a lock that is too old. 

Of course it would better if Knex would support something like that out of the box.

**Comment by elhigu** [comment]
Actually that is an interesting idea if triggers could be used to release migration lock in case where running migration fails unexpectedly without running code for releasing migrations. Maybe with trigger that is looking updates in some table having info about connections...

**Comment by mercmobily** [comment]
Thank you for raising this suggestion about adding a timestamp column to the migration lock table. This is indeed a valuable enhancement that would help address the "stuck lock" problem that occurs when a process crashes whilst holding the migration lock.

Currently, when a migration process crashes unexpectedly (due to OOM errors, container restarts, network issues, etc.), the lock remains set in the `knex_migrations_lock` table with `is_locked = 1`. This requires manual intervention using `knex migrate:unlock` to release the lock, which isn't ideal for automated deployments.

Looking at the current implementation in `lib/migrations/migrate/table-creator.js:45-52`, the lock table structure is quite minimal:

```javascript
function _createMigrationLockTable(tableName, schemaName, trxOrKnex) {
  return getSchemaBuilder(trxOrKnex, schemaName).createTable(
    tableName,
    function (t) {
      t.increments('index').primary();
      t.integer('is_locked');
    }
  );
}
```

**Proposed solution**

Adding timestamp columns would enable automatic lock timeout detection. The lock table could be enhanced with:
- `locked_at` - timestamp when lock was acquired
- `locked_by` - optional identifier of the process/host that acquired the lock

This would allow applications to implement their own timeout logic or for Knex to provide a built-in timeout mechanism. The workaround suggested by @electrotype using triggers is clever and shows this can work in practice.

Other migration tools like Flyway and Liquibase have similar timeout mechanisms. Flyway, for instance, has a `lockRetryCount` configuration and considers locks stale after a certain period.

**Implementation considerations**

1. **Backwards compatibility**: We'd need to handle existing lock tables gracefully, perhaps by adding columns if they don't exist
2. **Cross-database support**: Timestamp handling varies across database drivers (PostgreSQL, MySQL, SQLite, MSSQL, Oracle)
3. **Configuration**: Should provide a configurable timeout period (e.g., `migrationLockTimeout: 600000` for 10 minutes)
4. **Lock acquisition**: The `_lockMigrations` method would need to check both `is_locked` and the timestamp to determine if a lock is stale

This enhancement would make Knex migrations more robust in production environments, particularly in containerised deployments where processes can be terminated unexpectedly.

**Related issues**

- #5086 - Users experiencing stuck locks requiring manual intervention
- #3538 - Concurrent lock creation safety issues
- #4495 - Migration running twice in multi-server environments
- #1950 - Concurrent migrations failing instead of waiting
- #3780 - Migration lock issues requiring manual unlock
- #1217 - Questions about the purpose of knex_migration_lock table
